### ISSUE TITLE
Repo rules fail to extract unicode archives due to latin-1 hack

### ISSUE THREAD

**Issue by jvolkman** [bazelbuild/bazel#12986]
### Description of the problem / feature request:

`rules_go` currently [fails on some machines](https://github.com/bazelbuild/rules_go/issues/2771) due to some [unicode characters](https://github.com/golang/go/tree/master/test/fixedbugs/issue27836.dir) included in filenames within the Go source archive - specifically the character [Ä](https://www.compart.com/en/unicode/U+00C4). For Linux and macOS, Go archives are distributed as [`tar.gz`](https://golang.org/dl/) files with [pax headers](https://en.wikipedia.org/wiki/Tar_(computing)#POSIX.1-2001/pax) in the tar files.

Affected systems include:
* ZFS volumes with the `utf8only` option. This is Ubuntu's default when choosing ZFS at install time.
* macOS (HFS+ and APFS require UTF-8 filenames)

Bazel uses Apache Commons Compress to extract tar archives. For most tar files, Commons Compress defers to the encoding specified by the JVM's `-Dfile.encoding` param, or the platform default. With ISO-8859-1 - Bazel's preference - UTF-8 encoded filename bytes in tar files basically pass through verbatim when extracted and everything works.

But when the tar entry has a pax `path` header, the path name is [always decoded as UTF-8](https://github.com/apache/commons-compress/blob/rel/1.19/src/main/java/org/apache/commons/compress/archivers/tar/TarArchiveInputStream.java#L516-L517). 

The character Ä in its [composed](https://en.wikipedia.org/wiki/Unicode_equivalence) form is unicode character `U+00C4`. In Java's internal UTF-16, it's simply represented as `0xC4`. But encoded as UTF-8 it becomes the multi-byte sequence `0xC3 0x84` since `0xC4` as a single byte is not a valid UTF-8 value.

When Commons Compress parses a pax-formatted tar file with a filename containing Ä as the `0xC3 0x84` UTF-8 string, the resulting Java string contains the value `0xC4` after decoding. But this value is never re-encoded as UTF-8 when creating the file on the filesystem. Instead, Bazel uses the Java char values [verbatim as long as they're < 0xff](https://github.com/bazelbuild/bazel/blob/master/src/main/native/latin1_jni_path.cc#L84). An attempt to create a filename containing `0xC4` on a filesystem that requires UTF-8 filenames will fail.

But `rules_go` doesn't currently fail on macOS systems despite them also requiring UTF-8 filenames. This is because the `darwin` archives use decomposed representations of unicode characters. OS X has a history of preferring the decomposed forms over composed.

So instead of Ä being `U+00C4` ("Latin Capital Letter A with Diaeresis"), it's `U+0041` (just capital A) followed by `U+0308` ("Combining Diaeresis"). Encoded in UTF-8 as seen in the macOS golang tarballs, the byte string is `0x41 0xCC 0x88`. Decoded to a Java string (16-bit chars) it's `0x0041 0x0308`. Coincidentally I presume, Bazel is able to extract this decomposed form on UTF-8 filesystems because it [ignores](https://github.com/bazelbuild/bazel/blob/master/src/main/native/latin1_jni_path.cc#L84) the diaeresis and replaces it with a literal `'?'` character. So instead of `Äfoo.go` as is contained in the Go source archive, Bazel writes `A?foo.go` on macOS.

Like Linux on ZFS, Bazel fails to extract the linux archive on macOS as reported [here](https://github.com/bazelbuild/rules_go/issues/2773).

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

`test.bzl`
```python
def _tar_round_trip_impl(ctx):
    ctx.file("Äfoo.txt", "boo!\n")
    ctx.execute(["tar", "--format=" + ctx.attr.format, "-czvf", "file.tar.gz", "Äfoo.txt"])
    ctx.extract("file.tar.gz", "out")
    ctx.file("BUILD.bazel", 'exports_files(["Äfoo.txt", "out/Äfoo.txt", "file.tar.gz"])', legacy_utf8=False)

tar_round_trip = repository_rule(
    implementation = _tar_round_trip_impl,
    attrs = {
	"format": attr.string(
            mandatory = True,
        ),
    },
)
```

`WORKSPACE`
```python
load("//:test.bzl", "tar_round_trip")

tar_round_trip(
    name = "non_pax",
    format = "ustar",  # supported by both macos BSD tar and linux GNU tar
)

tar_round_trip(
    name = "pax",
    format = "pax",
)
```

`BUILD`
```python
genrule(
    name = "non_pax_test",
    srcs = ["@non_pax//:out/Äfoo.txt"],
    outs = ["non_pax.txt"],
    cmd = """
        cp $(location @non_pax//:out/Äfoo.txt) "$@"
    """,
)

genrule(
    name = "pax_test",
    srcs = ["@pax//:out/Äfoo.txt"],
    outs = ["pax.txt"],
    cmd = """
	cp $(location @pax//:out/Äfoo.txt) "$@"
    """,
)
```

```shell
# Works
bazel build //:non_pax_test

# Fails, either due to not being able to write the file (utf8 filesystem),
# or because the written filename is mangled.
bazel build //:pax_test
```

### What operating system are you running Bazel on?
Mac OS 10.15.7
Ubuntu 20.04 with a ZFS root partition with `utf8only` enabled (the default for Ubuntu's ZFS support).

### What's the output of `bazel info release`?
`release 4.0.0`

###  Have you found anything relevant by searching the web?
https://github.com/bazelbuild/rules_go/issues/2771
https://github.com/bazelbuild/bazel/issues/374 - pretty generic issue regarding filename characters.
https://github.com/bazelbuild/bazel/issues/7055 - an issue with the same problematic file in the Go archive, but targeted at Darwin only.

**Comment by illicitonion** [comment]
I happened to do some debugging of this yesterday - the probematic call chain is that https://github.com/bazelbuild/bazel/blob/2910d3d860044ef5fb0989306bafebc1c5545e42/src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java#L104 ends up calling into https://github.com/bazelbuild/bazel/blob/2910d3d860044ef5fb0989306bafebc1c5545e42/src/main/native/unix_jni.cc#L1045-L1058

Ideally we could use the raw buffer of bytes from the tar header to pass to `open`, rather than round-tripping via a String, but I'm not sure how to holistically go about that...

**Comment by aiuto** [comment]
cc @tetromino   This is a nice summary of the MacOS file handling strangeness I was talking about.

**Comment by brandjon** [comment]
Addressing the unicode handling situation for strings (replacing latin-1 with utf-8) is on the schedule for this year, perhaps in Q2.

**Comment by QIvan** [comment]
hi! Here is [a victim](https://github.com/envoyproxy/envoy/issues/16065) of this bug (I'm using Ubuntu 20.04 with ZFS and, seems like, it was set up with an `utf8only` flag, according to the original description). 
I tried to investigate the problem yesterday and copy some parts of Bazel's `CompressedTarFunction.java` to a small snippet. And, probably, I've found a workaround for this issue with just replacing `filePath.getOutputStream()` to `new FileOutputStream(filePath.getPathFile())`
So this code works for me https://gist.github.com/QIvan/bf88d152c31a35eccf162845ce05c455

**Comment by jvolkman** [comment]
@QIvan per my understanding this breaks other cases in which an archive contains a filename as a string of non-UTF-8 compatible bytes and the destination filesystem will accept it. #7757 has more context.

**Comment by brandjon** [comment]
Update: We haven't had bandwidth to work on eliminating the latin-1 hack, but may reconsider after 2023 Q1.

**Comment by aiuto** [comment]
I started a playground to observe some actual behaviors. We'll need to know how macos behaves when it sees NFC style names.
https://github.com/aiuto/bazel_samples/tree/main/utf8

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 6.4.0 RC2](https://releases.bazel.build/6.4.0/rc2/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
